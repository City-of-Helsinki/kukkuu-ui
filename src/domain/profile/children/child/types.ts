import type { UpdateChildMutationInput } from '../../../api/generatedTypes/graphql';

export type ChildDetailEditModalPayload = Omit<UpdateChildMutationInput, 'id'>;
