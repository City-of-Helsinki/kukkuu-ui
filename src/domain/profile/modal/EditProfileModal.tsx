import * as React from 'react';
import { Formik, FormikProps, FormikErrors, Form, FormikHelpers } from 'formik';
import { useTranslation } from 'react-i18next';
import { useMutation } from '@apollo/client';
import { toast } from 'react-toastify';
import * as Sentry from '@sentry/browser';
import { useMatomo } from '@jonkoops/matomo-tracker-react';
import * as yup from 'yup';

import styles from './editProfileModal.module.scss';
import {
  MyProfile,
  LanguagesSpokenAtHomeNode,
} from '../types/ProfileQueryTypes';
import Modal from '../../../common/components/modal/Modal';
import { SUPPORT_LANGUAGES } from '../../../common/translation/TranslationConstants';
import profileQuery from '../queries/ProfileQuery';
import updateMyProfileMutation from '../mutations/updateMyProfileMutation';
import {
  UpdateMyEmailMutationDocument,
  type RequestEmailUpdateTokenMutation,
  type UpdateMyEmailMutation,
  type UpdateMyProfileMutation,
  type UpdateMyProfileMutationInput,
  UpdateMyEmailMutationInput,
} from '../../api/generatedTypes/graphql';
import adultIcon from '../../../assets/icons/svg/adultFaceHappy.svg';
import NavigationConfirm from '../../../common/components/confirm/NavigationConfirm';
import FormikDropdown from '../../../common/components/formikWrappers/FormikDropdown';
import Button from '../../../common/components/button/Button';
import FormikTextInput from '../../../common/components/formikWrappers/FormikTextInput';
import LanguagesCombobox from '../../languages/LanguagesCombobox';
import RelayList from '../../api/relayList';
import requestEmailUpdateTokenMutation from '../mutations/requestEmailUpdateTokenMutation';
import LoadingSpinner from '../../../common/components/spinner/LoadingSpinner';

const schema = yup.object().shape({
  firstName: yup
    .string()
    .required('validation.general.required')
    .max(255, 'validation.maxLength'),
  lastName: yup
    .string()
    .required('validation.general.required')
    .max(255, 'validation.maxLength'),
  phoneNumber: yup
    .string()
    .required('validation.general.required')
    .max(255, 'validation.maxLength'),
  language: yup
    .string()
    .required('validation.general.required')
    .max(255, 'validation.maxLength'),
});

const emailChangeSchema = yup.object().shape({
  email: yup
    .string()
    .email('registration.form.guardian.email.input.error')
    .required('validation.general.required'),
  verificationToken: yup
    .string()
    .required('validation.general.required')
    .max(20, 'validation.maxLength'),
});

const LanguagesSpokenAtHome = RelayList<LanguagesSpokenAtHomeNode>();

function getInitialValues({
  languagesSpokenAtHome,
  ...rest
}: MyProfile): UpdateMyProfileMutationInput {
  return {
    ...rest,
    languagesSpokenAtHome: LanguagesSpokenAtHome(
      languagesSpokenAtHome
    ).items.map((language) => language.id),
  };
}

interface EditProfileModalProps {
  initialValues: MyProfile;
  isOpen: boolean;
  setIsOpen: (value: boolean) => void;
}

const EditProfileModal: React.FunctionComponent<EditProfileModalProps> = ({
  initialValues,
  isOpen,
  setIsOpen,
}) => {
  const { t } = useTranslation();

  const [isFilling, setFormIsFilling] = React.useState(false);

  return (
    <div>
      {isOpen && (
        <NavigationConfirm
          warningMessage={t('common.form.leave.warning.text')}
          isHalfFilling={isFilling}
        />
      )}
      <Modal
        setFormIsFilling={setFormIsFilling}
        label={t('registration.form.guardian.info.heading')}
        isOpen={isOpen}
        icon={adultIcon}
        toggleModal={(value: boolean) => {
          setIsOpen(value);
        }}
      >
        <EditMyEmailForm initialValues={initialValues} />
        <hr className={styles.separator} />
        <EditMyProfileForm
          initialValues={initialValues}
          setIsOpen={setIsOpen}
        />
      </Modal>
    </div>
  );
};

function EditMyEmailForm({
  initialValues: { email: originalEmail },
}: Pick<EditProfileModalProps, 'initialValues'>) {
  const { t } = useTranslation();

  const [isChangingEmail, setIsChangingEmail] = React.useState(false);
  const [updateMyEmail] = useMutation<UpdateMyEmailMutation>(
    UpdateMyEmailMutationDocument
  );

  const onSubmit = async (
    { email, verificationToken }: UpdateMyEmailMutationInput,
    { setFieldValue, setFieldError }: FormikHelpers<UpdateMyEmailMutationInput>
  ) => {
    try {
      await updateMyEmail({
        variables: {
          input: {
            email,
            verificationToken,
          },
        },
      });
      setIsChangingEmail(false);
      setFieldValue('verificationToken', '');
      toast.success(t('registration.submitMutation.successfulMessage'));
    } catch (error: any) {
      if (error.message.includes('token is invalid')) {
        setFieldError(
          'verificationToken',
          t('registration.form.guardian.email.verificationToken.errors.invalid')
        );
        toast.error(t('registration.form.guardian.email.errors.unsuccessful'));
      } else {
        toast.error(t('registration.submitMutation.errorMessage'));
        Sentry.captureException(error);
      }
    }
  };

  const validate = ({ email }: UpdateMyEmailMutationInput) => {
    if (!isChangingEmail) {
      setIsChangingEmail(true);
    }

    const errors: FormikErrors<UpdateMyEmailMutationInput> = {};

    if (email === originalEmail) {
      return {
        ...errors,
        email: 'registration.form.guardian.email.input.notChanged',
      };
    }
    return errors;
  };

  // Focus on email field when the email changing form is opened
  React.useEffect(() => {
    if (isChangingEmail) {
      (document.querySelector(`#email`) as HTMLInputElement)?.focus();
    }
  }, [isChangingEmail]);

  return (
    <Formik
      initialValues={{
        email: originalEmail,
        verificationToken: '',
      }}
      onSubmit={onSubmit}
      validate={validate}
      validationSchema={emailChangeSchema}
    >
      {({ isSubmitting }: FormikProps<UpdateMyEmailMutationInput>) => (
        <Form>
          <FormikTextInput
            className={styles.formField}
            id="email"
            name="email"
            disabled={!isChangingEmail}
            label={t('registration.form.guardian.email.input.label')}
          />
          {!isChangingEmail && (
            <Button
              className={styles.emailChangeButton}
              variant="supplementary"
              onClick={() => {
                setIsChangingEmail(true);
              }}
            >
              {t('registration.form.guardian.email.change.button')}
            </Button>
          )}
          {isChangingEmail && (
            <>
              <div className={styles.emailVerificationTokenFieldWrapper}>
                <FormikTextInput
                  className={styles.formField}
                  id="verificationToken"
                  name="verificationToken"
                  label={t(
                    'registration.form.guardian.email.verificationToken.input.label'
                  )}
                  placeholder={t(
                    'registration.form.guardian.email.verificationToken.input.placeholder'
                  )}
                  helperText={t(
                    'registration.form.guardian.email.verificationToken.input.helper'
                  )}
                  autoComplete="off"
                />
                <RequestVerificationTokenButton />
              </div>

              <div className={styles.buttonsWrapper}>
                <Button
                  onClick={() => setIsChangingEmail(false)}
                  variant="secondary"
                  type="button"
                  className={styles.cancelButton}
                >
                  {t('common.modal.cancel.text')}
                </Button>

                <Button
                  type="submit"
                  className={styles.submitButton}
                  disabled={isSubmitting}
                >
                  {t('common.modal.save.textWithSubject', {
                    subject: t('registration.form.guardian.email.input.label'),
                  })}
                </Button>
              </div>
            </>
          )}
        </Form>
      )}
    </Formik>
  );
}

function EditMyProfileForm({
  initialValues,
  setIsOpen,
}: Pick<EditProfileModalProps, 'initialValues' | 'setIsOpen'>) {
  const { t } = useTranslation();
  const { trackEvent } = useMatomo();
  const [isFilling, setFormIsFilling] = React.useState(false);
  const [updateMyProfile] = useMutation<UpdateMyProfileMutation>(
    updateMyProfileMutation,
    {
      refetchQueries: [{ query: profileQuery }],
    }
  );

  const onSubmit = async (payload: UpdateMyProfileMutationInput) => {
    setFormIsFilling(false);
    try {
      await updateMyProfile({
        variables: {
          input: {
            firstName: payload.firstName,
            lastName: payload.lastName,
            phoneNumber: payload.phoneNumber,
            language: payload.language,
            languagesSpokenAtHome: payload.languagesSpokenAtHome,
          },
        },
      });

      trackEvent({ category: 'action', action: 'Edit profile' });
      setIsOpen(false);
    } catch (error) {
      toast.error(t('registration.submitMutation.errorMessage'));
      Sentry.captureException(error);
    }
  };

  const validate = () => {
    if (!isFilling) {
      setFormIsFilling(true);
    }

    const errors: FormikErrors<UpdateMyProfileMutationInput> = {};
    return errors;
  };

  return (
    <Formik
      initialValues={getInitialValues(initialValues)}
      onSubmit={onSubmit}
      validate={validate}
      validationSchema={schema}
    >
      {({ isSubmitting }: FormikProps<UpdateMyProfileMutationInput>) => (
        <Form id="editProfileForm">
          <FormikTextInput
            className={styles.formField}
            id="phoneNumber"
            name="phoneNumber"
            type="tel"
            minLength={5}
            maxLength={255}
            required={true}
            label={t('registration.form.guardian.phoneNumber.input.label')}
            placeholder={t(
              'registration.form.guardian.phoneNumber.input.placeholder'
            )}
          />
          <div className={styles.profileName}>
            <FormikTextInput
              className={styles.formField}
              type="text"
              required={true}
              id="firstName"
              name="firstName"
              label={t('registration.form.guardian.firstName.input.label')}
              placeholder={t(
                'registration.form.guardian.firstName.input.placeholder'
              )}
            />
            <FormikTextInput
              className={styles.formField}
              type="text"
              required={true}
              id="lastName"
              name="lastName"
              label={t('registration.form.guardian.lastName.input.label')}
              placeholder={t(
                'registration.form.guardian.lastName.input.placeholder'
              )}
            />
          </div>
          <FormikDropdown
            className={styles.formField}
            id="language"
            name="language"
            label={t('registration.form.guardian.language.input.label')}
            required={true}
            options={[
              {
                label: t('common.language.en'),
                value: SUPPORT_LANGUAGES.EN.toUpperCase(),
              },
              {
                label: t('common.language.fi'),
                value: SUPPORT_LANGUAGES.FI.toUpperCase(),
              },
              {
                label: t('common.language.sv'),
                value: SUPPORT_LANGUAGES.SV.toUpperCase(),
              },
            ]}
            placeholder={t(
              'registration.form.guardian.language.input.placeholder'
            )}
          />
          <LanguagesCombobox
            helper={t(
              'registration.form.child.languagesSpokenAtHome.input.helper'
            )}
            id="languagesSpokenAtHome"
            label={t(
              'registration.form.child.languagesSpokenAtHome.input.label'
            )}
            name="languagesSpokenAtHome"
            // Block escape from closing the modal
            catchEscapeKey
          />
          <div className={styles.buttonsWrapper}>
            <Button
              onClick={() => {
                setIsOpen(false);
              }}
              variant="secondary"
              type="button"
              className={styles.cancelButton}
            >
              {t('common.modal.cancel.text')}
            </Button>

            <Button
              type="submit"
              className={styles.submitButton}
              disabled={isSubmitting}
            >
              {t('common.modal.save.text')}
            </Button>
          </div>
        </Form>
      )}
    </Formik>
  );
}

function RequestVerificationTokenButton() {
  const { t } = useTranslation();
  const [requestUpdateEmailVerificationToken, { loading }] =
    useMutation<RequestEmailUpdateTokenMutation>(
      requestEmailUpdateTokenMutation
    );

  const { seconds: coolDownSeconds, reset: resetCoolDown } = useCoolDown(
    60,
    false
  );

  const isDisabled = Boolean(loading || coolDownSeconds);

  const handleOnClick = async () => {
    try {
      await requestUpdateEmailVerificationToken();
      resetCoolDown();
      toast.success(
        t(
          'registration.form.guardian.email.verificationToken.request.successful'
        )
      );
    } catch (error) {
      toast.error(
        t('registration.form.guardian.email.verificationToken.request.error')
      );
      Sentry.captureException(error);
    }
  };

  return (
    <LoadingSpinner isLoading={loading}>
      <Button
        variant="supplementary"
        onClick={handleOnClick}
        className={styles.requestVerificationTokenButton}
        disabled={isDisabled}
      >
        {coolDownSeconds
          ? coolDownSeconds
          : t(
              'registration.form.guardian.email.verificationToken.request.button'
            )}
      </Button>
    </LoadingSpinner>
  );
}

function useCoolDown(timerSeconds: number, start = false) {
  const [coolDownSeconds, setCoolDownSeconds] = React.useState(
    start ? timerSeconds : 0
  );

  React.useEffect(() => {
    const interval = setInterval(() => {
      if (coolDownSeconds > 0) {
        setCoolDownSeconds(coolDownSeconds - 1);
      }
    }, 1000);
    return () => clearInterval(interval);
  }, [coolDownSeconds]);

  const resetCoolDown = () => {
    setCoolDownSeconds(timerSeconds);
  };

  return { seconds: coolDownSeconds, setCoolDownSeconds, reset: resetCoolDown };
}

export default EditProfileModal;
