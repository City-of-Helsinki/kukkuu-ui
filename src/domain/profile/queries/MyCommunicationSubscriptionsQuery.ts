import { gql } from '@apollo/client';

const myCommunicationSubscriptionsQuery = gql`
  query MyCommunicationSubscriptions($authToken: String) {
    myCommunicationSubscriptions(authToken: $authToken) {
      firstName
      lastName
      language
      hasAcceptedCommunication
    }
  }
`;
export default myCommunicationSubscriptionsQuery;
