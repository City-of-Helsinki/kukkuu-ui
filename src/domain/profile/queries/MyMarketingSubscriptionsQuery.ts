import { gql } from '@apollo/client';

const myMarketingSubscriptionsQuery = gql`
  query MyMarketingSubscriptions($authToken: String) {
    myMarketingSubscriptions(authToken: $authToken) {
      firstName
      lastName
      language
      hasAcceptedMarketing
    }
  }
`;
export default myMarketingSubscriptionsQuery;
