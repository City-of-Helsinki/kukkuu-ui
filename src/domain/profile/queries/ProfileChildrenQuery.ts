import gql from 'graphql-tag';

const profileChildrenQuery = gql`
  query profileChildrenQuery {
    myProfile {
      id
      children {
        edges {
          node {
            id
          }
        }
      }
    }
  }
`;
export default profileChildrenQuery;
