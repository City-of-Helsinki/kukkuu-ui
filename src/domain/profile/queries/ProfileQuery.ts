import gql from 'graphql-tag';

const profileQuery = gql`
  query profileQuery {
    myProfile {
      id
      firstName
      lastName
      email
      phoneNumber
      language
      children {
        edges {
          node {
            id
            firstName
            lastName
            birthdate
            postalCode
            project {
              id
              name
              year
            }
            relationships {
              edges {
                node {
                  id
                  type
                }
              }
            }
            availableEvents {
              edges {
                node {
                  id
                  name
                  duration
                  participantsPerInvite
                }
              }
            }
            occurrences {
              edges {
                node {
                  id
                  event {
                    id
                    name
                  }
                }
              }
            }
          }
        }
      }
      languagesSpokenAtHome {
        edges {
          node {
            id
          }
        }
      }
    }
  }
`;
export default profileQuery;
