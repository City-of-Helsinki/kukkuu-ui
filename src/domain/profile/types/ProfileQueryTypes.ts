import { ProfileQuery } from '../../api/generatedTypes/graphql';

export type MyProfile = NonNullable<ProfileQuery['myProfile']>;

export type MyProfileChildren = MyProfile['children'];

export type MyProfileChild = NonNullable<
  NonNullable<MyProfileChildren['edges'][number]>['node']
>;

export type MyProfileEnrolment = NonNullable<
  NonNullable<MyProfileChild['enrolments']['edges'][number]>['node']
>;

export type LanguagesSpokenAtHomeNode = NonNullable<
  MyProfile['languagesSpokenAtHome']['edges'][number]
>['node'];

export type Project = MyProfileChild['project'];
