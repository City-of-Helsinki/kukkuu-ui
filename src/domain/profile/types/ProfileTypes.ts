export default interface ProfileValues {
  firstName: string;
}
