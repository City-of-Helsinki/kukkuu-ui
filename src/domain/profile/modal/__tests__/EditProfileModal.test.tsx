/* eslint-disable testing-library/no-node-access */
import React from 'react';
import { screen } from '@testing-library/react';
import { MockedResponse } from '@apollo/client/testing';

import EditProfileModal from '../EditProfileModal';
import { MyProfile } from '../../types/ProfileQueryTypes';
import { Language } from '../../../api/generatedTypes/graphql';
import {
  render,
  fireEvent,
  waitFor,
} from '../../../../common/test/testingLibraryUtils';
import initModal from '../../../../common/test/initModal';
import { languagesQueryResponse } from '../../../app/footer/__mocks__/languagesMock';
import { languagesQuery } from '../../../languages/queries/LanguageQueries';

const initialValues: MyProfile = {
  id: 'yuiop',
  firstName: 'asdf',
  lastName: 'fdsa',
  phoneNumber: '0904422233',
  email: 'email@example.com',
  language: Language.Sv,
  hasAcceptedCommunication: false,
  children: {
    edges: [],
  },
  languagesSpokenAtHome: {
    edges: [],
  },
};

const defaultProps = {
  initialValues,
  isOpen: true,
  setIsOpen: () => {},
};

const formData = {
  phoneNumber: '000000000',
  firstName: 'George',
  lastName: 'Lopez',
};

const languagesMock: MockedResponse = {
  request: {
    query: languagesQuery,
    variables: {},
  },
  result: { ...languagesQueryResponse },
};

const mocks: MockedResponse[] = [languagesMock];

const getHdsSelect = (elements: HTMLElement[]) => {
  return elements[0].parentElement;
};

const selectOption = (
  selectWrapper: HTMLElement | null,
  optionLabel: string
) => {
  const button = selectWrapper?.querySelector('button');
  const input = selectWrapper?.querySelector('input');
  const controller = button || input;

  if (!controller) {
    throw Error('Could not find controlling element');
  }

  fireEvent.click(controller);

  const options = selectWrapper?.querySelectorAll('[role="option"]');
  const optionToBeSelected = Array.from(options || []).find((element) => {
    return element.textContent === optionLabel;
  });

  if (!optionToBeSelected) {
    throw Error(`Option with label "${optionLabel}" could not be found`);
  }

  fireEvent.click(optionToBeSelected);
};

it('renders snapshot correctly', () => {
  initModal();
  const { container } = render(
    <EditProfileModal
      initialValues={initialValues}
      isOpen={true}
      setIsOpen={vi.fn()}
    />,
    mocks
  );
  expect(container).toMatchSnapshot();
});

it('should allow all fields to be filled', async () => {
  initModal();
  render(<EditProfileModal {...defaultProps} />, mocks);

  fireEvent.change(screen.getByRole('textbox', { name: 'Puhelinnumero *' }), {
    target: { value: formData.phoneNumber },
  });
  fireEvent.change(screen.getByRole('textbox', { name: 'Etunimi *' }), {
    target: { value: formData.firstName },
  });
  fireEvent.change(screen.getByRole('textbox', { name: 'Sukunimi *' }), {
    target: {
      value: formData.lastName,
    },
  });
  selectOption(
    getHdsSelect(screen.getAllByRole('button', { name: /asiointikieli \*/i })),
    'Suomi'
  );

  // Because the component submits with a GraphQL hook we don't look
  // at the submit function, but instead just try and verify that we can
  // find all the values we set during this test.
  Object.values(formData).forEach((value) => {
    expect(screen.queryByDisplayValue(value)).not.toEqual(null);
  });
  // Handle select as a special case because it has no input
  await waitFor(() => {
    expect(
      getHdsSelect(
        screen.getAllByRole('button', { name: /asiointikieli \*/i })
      )?.querySelector('button')?.textContent
    ).toEqual('Suomi');
  });
}, 10_000);
