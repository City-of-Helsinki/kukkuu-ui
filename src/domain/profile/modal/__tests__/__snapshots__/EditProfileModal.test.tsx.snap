// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders snapshot correctly 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <MockedProvider
    addTypename={false}
  >
    <HelmetProvider
      context={Object {}}
    >
      <Router
        history={
          Object {
            "action": "POP",
            "block": [Function],
            "createHref": [Function],
            "go": [Function],
            "goBack": [Function],
            "goForward": [Function],
            "length": 1,
            "listen": [Function],
            "location": Object {
              "hash": "",
              "pathname": "/",
              "search": "",
              "state": undefined,
            },
            "push": [Function],
            "replace": [Function],
          }
        }
      >
        <MockedProvider
          addTypename={true}
        >
          <EditProfileModal
            initialValues={
              Object {
                "children": Object {
                  "edges": Array [],
                },
                "email": "email@example.com",
                "firstName": "asdf",
                "id": "yuiop",
                "language": "SV",
                "languagesSpokenAtHome": Object {
                  "edges": Array [],
                },
                "lastName": "fdsa",
                "phoneNumber": "0904422233",
              }
            }
            isOpen={true}
            setIsOpen={[MockFunction]}
          />
        </MockedProvider>
      </Router>
    </HelmetProvider>
  </MockedProvider>
</Provider>
`;
